File: 백준/단계 별로 풀어보기/기본_수학_2/9020_골드바흐의_추측.py
import sys
input = sys.stdin.readline

def solve():
    Q = [int(input()) for T in range(int(input()))]
    
    end = max(Q) + 1
    save = [True] * (end // 2)
    for i in range(3, int(end ** 0.5) + 1, 2):
        if save[i // 2]:
            k = i * i
            save[k // 2 :: i] = [False] * ((end - k - 1) // (2 * i) + 1)
    primes = {2 * i + 1 for i in range(1, end // 2) if save[i]}
    
    result = ""
    for each in Q:
        if each == 4: result += '2 2\n'; continue
        half = each // 2
        for i in range(not half & 1, half, 2):
            if half + i in primes and half - i in primes: result += f"{half - i} {half + i}\n"; break
    
    print(result)
    
if "__main__" == __name__: solve()

File: 백준/단계 별로 풀어보기/기본_수학_2/9020_골드바흐의_추측.cpp
#include <iostream>
#include <vector>
#include <string>
#include <algorithm>
#include <cmath>
#include <set>
using namespace std;

int main()
{
	int T, input;
	scanf("%d", &T);
	
	vector<int> Q;
	while (T--){
	    scanf("%d", &input);
	    Q.push_back(input);
	}
	
	int end = *max_element(Q.begin(), Q.end()) + 1;
	bool *save = new bool[end / 2];
	fill_n(save, end / 2, true);
	
	for (int i = 3; i < (int)sqrt(end) + 1; i += 2)
	{
		if (save[i / 2]){
			int k = i * i;
			for (int j = k / 2; j < end / 2; j += i) save[j] = false;
		}
	}
	
	set<int> primes;
	for (int i = 1; i < end / 2; i++){
		if (save[i] == true) primes.insert(2 * i + 1);
	}
	
	unsigned int half, start;
	for (unsigned int each : Q){
	    if (each == 4){
	        cout << "2 2\n";
	        continue;
	    }
	    half = each / 2;
	    if (!(half & 1)) start = 1;
	    else start = 0;
	    
	    for (int i = start; i < half; i += 2){
	        if (primes.count(half - i) == 1 and primes.count(half + i) == 1){
	            printf("%d %d\n", half - i, half + i);
	            break;
	        }
	    }
	}
}